src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractNonSubstitutableMemberArgumentMatcherAnalyzer.cs

```diff
@@ -84,27 +92,172 @@ private void AnalyzeArgLikeMethod(SyntaxNodeAnalysisContext syntaxNodeContext, T
+            if (_nonSubstitutableMemberAnalysis.Analyze(
+                syntaxNodeContext,
+                enclosingExpression,
+                enclosingExpressionSymbol).CanBeSubstituted != false)
+            {
+                return false;
+            }
+
+            syntaxNodeContext.TryReportDiagnostic(
+                Diagnostic.Create(
+                    DiagnosticDescriptorsProvider.NonSubstitutableMemberArgumentMatcherUsage,
+                    argInvocationExpression.GetLocation()),
+                enclosingExpressionSymbol);
+
+            return true;
+        }
+
+        private bool AnalyzeEnclosingExpressionCandidateSymbols(
+            SyntaxNodeAnalysisContext syntaxNodeContext,
+            TInvocationExpressionSyntax argInvocationExpression,
+            SyntaxNode enclosingExpression,
+            SymbolInfo enclosingExpressionSymbolInfo)
+        {
+            if (enclosingExpressionSymbolInfo.CandidateSymbols.Length == 0)
             {
                 var diagnostic = Diagnostic.Create(
                     DiagnosticDescriptorsProvider.NonSubstitutableMemberArgumentMatcherUsage,
                     argInvocationExpression.GetLocation());
 
-                syntaxNodeContext.TryReportDiagnostic(diagnostic, enclosingExpressionSymbol);
+                syntaxNodeContext.TryReportDiagnostic(diagnostic, null);
+                return true;
+            }
+
+            foreach (var candidateSymbol in enclosingExpressionSymbolInfo.CandidateSymbols)
+            {
+                if (_nonSubstitutableMemberAnalysis.Analyze(
+                    syntaxNodeContext,
+                    enclosingExpression,
+                    candidateSymbol).CanBeSubstituted == false)
+                {
+                    var diagnostic = Diagnostic.Create(
+                        DiagnosticDescriptorsProvider.NonSubstitutableMemberArgumentMatcherUsage,
+                        argInvocationExpression.GetLocation());
+
+                    syntaxNodeContext.TryReportDiagnostic(diagnostic, candidateSymbol);
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void AnalyzeAssignment(
+            SyntaxNodeAnalysisContext syntaxNodeContext,
+            TInvocationExpressionSyntax argInvocationExpression,
+            IMethodSymbol argInvocationExpressionSymbol,
+            IMemberReferenceOperation memberReferenceOperation)
+        {
+            if (memberReferenceOperation == null)
+            {
+               return;
+            }
+
+            var syntaxNode = memberReferenceOperation.Syntax;
+
+            if (IsWithinWhenLikeMethod(syntaxNodeContext, syntaxNode))
+            {
+                return;
+            }
+
+            if (argInvocationExpressionSymbol.IsArgDoLikeMethod())
+            {
+               return;
+            }
+
+            if (IsPrecededByReceivedLikeMethod(syntaxNodeContext, syntaxNode))
```

Comment:
@dtchepak Considering the fact that NS1004 will be reported for cases when (for instance) `Arg.Is` is used in following context
```c#
 substitute.Type = Arg.Is("sth");
```
instead of 
```
 substitute.Received().Type = Arg.Is("sth");
```
even if Type property is virtual, should we change default message
```
[NS1004] Argument matcher used with a non-virtual member of a class.
```
to sth else?
Documentation clearly states that 
`This violation can also be reported if an argument matcher is used for something other than setting return values, checking received calls, or configuring callbacks. See [NSubstitute's argument matcher documentation](https://nsubstitute.github.io/help/argument-matchers/#how_not_to_use_argument_matchers) for more information.` but maybe we should "promote" this message (or shorter one) to warning visible during the build?

Reply:
I think ideally this should be a different error ("5003 Misused argument matcher"?).

The non-virtual cases are very clear in terms of cause and fix. The misuse case is a lot more subtle.

Maybe can tackle that as a separate issue so this fix can be merged first?